src/main/java/org/mongojack/internal/stream/JacksonCodec.java

```diff
@@ -30,4 +49,197 @@ public void encode(BsonWriter writer, T value, EncoderContext encoderContext) {
+            }
+        }
+    }
+
+    private Optional<Field> getIdField(final Class<?> documentClass) {
+        Field[] fields = documentClass.getDeclaredFields();
+        Optional<Field> maybeField = Arrays.stream(fields)
+            .filter(field -> field.isAnnotationPresent(javax.persistence.Id.class) ||
+                field.isAnnotationPresent(org.mongojack.Id.class) ||
+                field.getName().equals("_id"))
+            .findFirst();
+        if (maybeField.isPresent()) {
+            return maybeField;
+        }
+
+        Class<?> superClass = documentClass.getSuperclass();
+        if (superClass != null && !Object.class.equals(superClass)) {
+            return getIdField(superClass);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<Method> getIdGetter(final Class<?> documentClass) {
+        Method[] methods = documentClass.getDeclaredMethods();
+        Optional<Method> maybeGetter = Arrays.stream(methods)
+            .filter(method -> method.getName().startsWith("get") &&
+                method.getParameterCount() == 0 &&
+                (method.isAnnotationPresent(javax.persistence.Id.class) ||
+                    method.isAnnotationPresent(org.mongojack.Id.class) ||
+                    method.getName().equals("get_id")))
+            .findFirst();
+        if (maybeGetter.isPresent()) {
+            return maybeGetter;
+        }
+
+        Class<?> superClass = documentClass.getSuperclass();
+        if (superClass != null && !Object.class.equals(superClass)) {
+            return getIdGetter(superClass);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<Method> getIdSetter(final Class<?> documentClass) {
+        Method[] methods = documentClass.getDeclaredMethods();
+        Optional<Method> maybeSetter = Arrays.stream(methods)
+            .filter(method -> method.getName().startsWith("set") &&
+                method.getParameterCount() == 1 &&
+                (method.isAnnotationPresent(javax.persistence.Id.class) ||
+                    method.isAnnotationPresent(org.mongojack.Id.class) ||
+                    method.getName().equals("set_id")))
+            .findFirst();
+        if (maybeSetter.isPresent()) {
+            return maybeSetter;
+        }
+
+        Class<?> superClass = documentClass.getSuperclass();
+        if (superClass != null && !Object.class.equals(superClass)) {
+            return getIdSetter(superClass);
+        }
+        return Optional.empty();
+    }
+
+    private Object extractValue(BsonObjectId value, Class<?> valueType) {
+        if (String.class.equals(valueType)) {
+            return value.asObjectId().getValue().toHexString();
+        } else if (ObjectId.class.equals(valueType)) {
+            return value.asObjectId().getValue();
+        }
+        throw new IllegalArgumentException("Unsupported ID type: " + value.getClass());
+    }
+
+    public static Object extractValueEx(BsonValue value) {

```

Comment:
Is this method used anywhere or can it be removed?

Reply:
Yes, it's called from JacksonMongoCollection